### DISCUSSION TITLE
How to make Bootstrap-vue Toast in an independent javascript file

### DISCUSSION THREAD

**Discussion by arunstop** [bootstrap-vue/bootstrap-vue#6257] [Q&A] [answered]
It's been a few days since I have been learning **vue.js**, I wanted to make a function that shows a Toast, reading from the official bootstrap-vue docs : https://bootstrap-vue.org/docs/components/toast#toasts-on-demand, I was able to make **toast via component instance injection** and from **custom component**, however, I want to make another custom toast via independent **javascript function /file** so I can call it dynamically.

Here's **Cart.vue** code which i used injection: 
```
<script>
export default {
  methods: {
    purchaseHandler() {
      this.$bvToast.toast('The item added to your cart', {
          title: 'Notification',
          variant: 'succes',
          solid: true
        })
    },
  },
};
</script>
```

However, I made a **custom-toast.js** as an independent js function/file and imported the plugin as the docs stated: 
```
import Vue from 'vue'
import { ToastPlugin } from 'bootstrap-vue'
Vue.use(ToastPlugin)

export default () => {
    this.$bvToast.toast('Toast body content', {
        title: 'Toast test',
        variant: 'succes',
        solid: true
      })
}
```

And I imported **custom-toast.js** to my **Cart.vue** file, now it looks this way : 
```
<script>
import customToast from "./custom-toast";
export default {
  methods: {
    purchaseHandler() {
      customToast()
  },
};
</script>
```

Now comes THE ERROR:

**[Vue warn]: Error in v-on handler: "TypeError: Cannot read property '$bvToast' of undefined"**

**TypeError: Cannot read property '$bvToast' of undefined**

I know this is super basic stuff, it thought I can solve it by finding the equivalent of **this.$bvToast** injection in the independent js file, but I couldn't find any, since my knowledge of Vue js is still small. I also tried to pass a parameter valued **this.$bvToast** from the .vue file to the .js file, but i don't want it this way, so the the real questions are: 

 1. What is the equivalent of **this.$bvToast** or any bootstrap-vue injection in an independent js file?
 2. How do I solve this problem? 

Thank you. Sorry for my bad English

**Comment by jacobmllr95** [comment, marked as answer]
@arunstop Check out this CodeSandbox: https://codesandbox.io/s/pensive-yonath-yrmkk

I would suggest creating a [mixin](https://vuejs.org/v2/guide/mixins.html) with a method to create the toast that you can reuse in multiple components.

**Comment by arunstop** [comment]
Thanks for the answer Jacob! This works perfectly fine, although I have made my own version of this via **Vue.prototype** and that leads to this question: In the real world scenario, which one is considered more "correct" or "proper", using **prototypes** or **mixins** ? Cheers

**Reply by jacobmllr95** [reply]
I would personally go with the mixin but in the end it depends on your exact usecase.
